### DISCUSSION TITLE
[IDEA] New docker-based dev tool

### DISCUSSION THREAD

**Discussion by terminalmage** [saltstack/salt#67354] [Feature Requests]
# Problem

Getting a development environment up and running can take awhile, especially if you are new to things like pyenv, virtualenvs, etc. Additionally, what if you are developing on a Fedora machine and need to work on CentOS-related code, or your dev box is Arch Linux and you need to work on Ubuntu-specific code?

Salt could use something to reduce development friction and make it easier to start hacking on Salt.

# Solution

A while back, I collaborated on a project called [barnacle](https://github.com/cachedout/barnacle), which was a collection of Dockerfiles to set up an image, install the Salt depchain, and configure the PYTHONPATH such that one could mount a clone of Salt into an instance of a Docker image and run Salt against the current code in your git clone.

I have taken the concepts from that project and designed a Dockerfile (using BuildKit-specific syntax) which can be used to create images for different Linux distros. The Dockerfile builds Python using pyenv, and then creates a virtualenv and installs the static requirements file used by Salt's CI pipelines to ensure that all of the necessary libraries are present.

The code and documentation can be found in [this repo](https://github.com/terminalmage/salt-docker).

**Comment by s0undt3ch** [comment]
First of all, nice solution!

Just a small question, instead of building Python, why not pick one of the nightly builds of salt(depending on the branch you want to work on) and run tests against that?

I'll answer that myself too, because of you're changing Salt, then the onedir won't be that useful....

However, why not use our tools command, and build the base Onedir and use that? (Looking at the workflows you can see how we do it in CI)

As for where it could land, depends on how big it might become. Let me punt on that.

**Comment by s0undt3ch** [comment]
Let me also bring @cmcmarrow into the topic since he's also been working on something similar.

**Comment by terminalmage** [comment]
@s0undt3ch and I spoke on Slack, but to bring some of that discussion into this GitHub issue, I'll include some comments below:

> First of all, nice solution!
> 
> Just a small question, instead of building Python, why not pick one of the nightly builds of salt(depending on the branch you want to work on) and run tests against that?
> 
> I'll answer that myself too, because of you're changing Salt, then the onedir won't be that useful....
> 

Yeah, this is designed to be able to develop on Salt while you're making changes. You develop using your preferred editor/IDE, and can immediately run the code on the specified platform.

> However, why not use our tools command, and build the base Onedir and use that? (Looking at the workflows you can see how we do it in CI)
> 

There's no facility in that tool right now to use Docker. My shell script frontend could be integrated into that tool, however.


We also spoke about the use of `platform.platform()`, and ways we could add additional code there to replace the usage of `platform.platform()` in our skipif conditions, without resulting in code-duplication. Changes there will need to precede any changes to the test suite.

**Comment by cmcmarrow** [comment]
@terminalmage here is my current work https://github.com/cmcmarrow/saltbox. It needs help and is by far not perfect. Here is an example of the saltbox API. Its only in python right now, need to move to yaml.

```
from saltbox.utils.file import FileBridge, FileData
from saltbox import saltbox
from saltbox.orchestration import Orchestration

APACHE = """
apache:
  service.running:
    - require:
      - pkg: apache"""

with Orchestration([(1, saltbox.SaltBoxBuilder("docker",
                                               "build",
                                               master_build=True,
                                               minion_build=True,
                                               ports={"4444", "4333"},
                                               state_files=[FileBridge(FileData("a.sls", APACHE), "a.sls")],
                                               salt="3005")),
                    (2, saltbox.SaltBoxBuilder("docker",
                                               "minion",
                                               minion_build=True,
                                               copies=[FileBridge(FileData("data.txt", "DATA MEOW"), "/data.txt")])),
                    (2, saltbox.SaltBoxBuilder("docker",
                                               "minion_old",
                                               minion_build=True,
                                               python_packages=("saltext.vmware",),
                                               salt="3004"))
                    ]) as d:
    print(list(d))
    print(d["build_master"].run("salt '*' test.ping").output)
    print(d["build_master"].run("salt '*' cmd.run 'ls /'").output)
    print(d["minion_old_1"].run("salt-call --local test.ping").output)
    print(d["build_master"].run("salt '*' state.apply a").output)
    print(d["build_master"].run("ls srv/salt").output)
```

Here we are making 5 minions:
*  2 are lastest release 
* 2 are 3004
* master is a minion as well at version 3005

Orchestration names the minions and connect them to the master. Then in the with block you have shell access to VMs.

If you would like to talk more and share ideas I think we should move to slack.
